src/mpid/ch4/netmod/ucx/ucx_pre.h

```diff
@@ -33,12 +38,23 @@ typedef struct {
     int handler_id;
     char *pack_buffer;
     ucp_dt_iov_t iov[2];
+    int cmpl_cntr;              /* am message may be sent in pieces,
+                                 * thus need track completion status */
 } MPIDI_UCX_am_request_t;
 
+typedef enum {
+    MPIDI_UCX_AM_PACK_INVALID,
+    MPIDI_UCX_AM_PACK_ONE,      /* all in one msg, if fit inside MPIDI_UCX_AM_BUF_SIZE */
+    MPIDI_UCX_AM_PACK_TWO_A,    /* two msgs, msg_hdr + am_hdr and payload */
+    MPIDI_UCX_AM_PACK_TWO_B,    /* two msgs, msg_hdr and am_hdr + payload, if am_hdr is too big */
```

Comment:
`PACK_TWO_B` may not be the case in MPICH. The only place that we might create a large am_hdr is when RMA needs to pass flattened datatype in am_hdr. But for that scenario, we already check if the total am_hdr size is smaller than the NM's header size limit (which is eager limit in UCX). If the am_hdr is to large for one eager message, we will just send flattened datatype as payload (PACK_TWO_A).

Reply:
I know. I am adding it for completeness. As a component, its functionality and correctness should try not to rely on (too much) usage assumptions. Otherwise, the component code cannot be maintained and reviewed stand-alone and the change in upper layer or lower layer cannot be easily achieved independently. I think a little flexibility, even when it is not needed for the current code, can ease the long term maintenance.